File: python/palu_factor.py
# PA=LU factorization algorithm

def palu_factor(A):
    '''
    Input:
      - A: a numpy coefficient square matrix
    Output:
      - A: LU factorized matrix with the upper trianguar part
           containing U and the lower trianguar part (diagonal
           excluded) containing L.
      - p: permutation vector
    '''
    # Get the matrix size
    n = A.shape[0]
    p = list(range(0, n))
    if n != A.shape[1]:
        raise Exception('The coefficient matrix should be square')
    for i in range(0,n-1):
        # Get the max row number
        maxrow = abs(A[i:n,i]).argmax()
        # align vector row with matrix row
        maxrow = maxrow + i
        maxval = A[maxrow,i]
        if maxval == 0:
            raise Exception('The matrix is not LU factorizable')
        if maxrow > i:
            # rows swap
            A[[i,maxrow]]=A[[maxrow,i]]
            #p[[i,maxrow]]=p[[maxrow,i]]
            p[i], p[maxrow] = p[maxrow], p[i]
        # computation of the i-th Gauss vector that will be
        # part of the L matrix
        A[i+1:,i] = A[i+1:,i]/A[i,i]
        # recomputation of the square sub-matrix defined under
        # the i-th row and after the i-th column
        A[i+1:,i+1:] = A[i+1:,i+1:] - A[i+1:,i]*A[i,i+1:]
    return (A,p)

File: octave/palu_factor.m
% PA=LU factoriazation algorithm
%
% input:
%   - A : the coefficient matrix
% output:
%   - A : the LU factorized matrix
%   - p: permutation vector
%
function [A,p] = palu_factor(A)
    n = length(A);
    % permutation indexes vector initialization
    p = 1:n;
    for i = 1:n-1
        [maxval,row] = max(abs(A(i:n,i)));
        % align vector row with matrix row
        row = row+i-1;
        % for each iteration check that maxval is non-zero
        if maxval == 0
            error('The matrix is singular');
        end
        if row > i
            % rows swap
            A([i,row],:) = A([row,i],:);
            % update the permutation vector
            p([i,row]) = p([row,i]);
        end
        % computation of the i-th Gauss vector that will be part of the L matrix
        A(i+1:n,i) = A(i+1:n,i)/A(i,i);
        % recomputation of the square sub-matrix defined under the i-th row and
        % after the i-th column
        A(i+1:n,i+1:n) = A(i+1:n,i+1:n) - A(i+1:n,i)*A(i,i+1:n);
    end
end
